#1012 RxRatpack#fork for parallel observables

tednaleid opened the pull request:
Implements `RxRatpack#fork` by leveraging the work by @alkemist that was just added.

I used the same name as was used for the `Promise#fork`, but there could be some discussion as to how it could be a little confusing how it differs from `forkEach`.  

I tried to highlight in the documentation and tests how `fork` modifies the execution that the upstream observable is run on and gets bound back to the original execution after the `fork`, whereas `forkEach` modifies where downstream gets executed on.  It's quite a bit similar to the RxJava `subscribeOn` vs `observeOn`.

I also modified the existing tests for `forkEach` to demonstrate that you don't need to use `compose` within groovy code.  That wasn't obvious to me when I was looking through the tests originally trying to understand how it worked.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/ratpack/ratpack/1014)
<!-- Reviewable:end -->

beckje01 commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[ratpack-rx/src/test/groovy/ratpack/rx/RxParallelSpec.groovy, line 19 \[r1\]](https://reviewable.io:443/reviews/ratpack/ratpack/1014#-KKmxVE1v9f0PVwOCSF4:-KKmxVE1v9f0PVwOCSF5:-1855219655) ([raw file](https://github.com/ratpack/ratpack/blob/4e3db681ed5184a8cf5087346063ae2e6ebabe51/ratpack-rx/src/test/groovy/ratpack/rx/RxParallelSpec.groovy#L19)):*
> ```Groovy
> package ratpack.rx
> 
> import ratpack.exec.Promise
> ```

Unused import and its stopping the build due to codenarc violation

---

*[ratpack-rx/src/test/groovy/ratpack/rx/RxParallelSpec.groovy, line 298 \[r1\]](https://reviewable.io:443/reviews/ratpack/ratpack/1014#-KKmxfGDWyi7jdQZAAqz:-KKmxfGDWyi7jdQZAAr-:747710510) ([raw file](https://github.com/ratpack/ratpack/blob/4e3db681ed5184a8cf5087346063ae2e6ebabe51/ratpack-rx/src/test/groovy/ratpack/rx/RxParallelSpec.groovy#L298)):*
> ```Groovy
>     Integer result = harness.yieldSingle {
>       sequence
>         .doOnNext { if (it == 3) throw new RuntimeException("3!") }
> ```

Codenarc rules enforce all if statements to have braces so no single line if here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/ratpack/ratpack/1014)*
<!-- Sent from Reviewable.io -->

tednaleid commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[ratpack-rx/src/test/groovy/ratpack/rx/RxParallelSpec.groovy, line 19 \[r1\]](https://reviewable.io:443/reviews/ratpack/ratpack/1014#-KKmxVE1v9f0PVwOCSF4:-KKn7iWdeTMqWYTECZuV:-499146009) ([raw file](https://github.com/ratpack/ratpack/blob/4e3db681ed5184a8cf5087346063ae2e6ebabe51/ratpack-rx/src/test/groovy/ratpack/rx/RxParallelSpec.groovy#L19)):*
<details><summary><i>Previously, beckje01 (Jeff Beck) wrote…</i></summary>
> Unused import and its stopping the build due to codenarc violation
</details>
Done.

---

*[ratpack-rx/src/test/groovy/ratpack/rx/RxParallelSpec.groovy, line 298 \[r1\]](https://reviewable.io:443/reviews/ratpack/ratpack/1014#-KKmxfGDWyi7jdQZAAqz:-KKn7j9u_19tiv-c0dr8:-499146009) ([raw file](https://github.com/ratpack/ratpack/blob/4e3db681ed5184a8cf5087346063ae2e6ebabe51/ratpack-rx/src/test/groovy/ratpack/rx/RxParallelSpec.groovy#L298)):*
<details><summary><i>Previously, beckje01 (Jeff Beck) wrote…</i></summary>
> Codenarc rules enforce all if statements to have braces so no single line if here.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/ratpack/ratpack/1014)*
<!-- Sent from Reviewable.io -->

tednaleid commented:
build [is failing](https://snap-ci.com/ratpack/ratpack/pull/1014/logs/defaultPipeline/3/Test) based on manual test failures that didn't fail for me locally and I'm pretty sure I didn't change...

```
:ratpack-manual:test

ratpack.manual.JavadocCodeSnippetTests > ratpack.exec.util.Batch:35 FAILED
    ratpack.manual.snippets.executer.CompileException: java.lang.AssertionError: ERROR: package ratpack.exec.batch does not exist (2:27)

        Caused by:
        java.lang.AssertionError: ERROR: package ratpack.exec.batch does not exist (2:27)

ratpack.manual.JavadocCodeSnippetTests > ratpack.exec.util.ParallelBatch:105 FAILED
    ratpack.manual.snippets.executer.CompileException: java.lang.AssertionError: ERROR: package ratpack.exec.batch does not exist (3:27)

        Caused by:
        java.lang.AssertionError: ERROR: package ratpack.exec.batch does not exist (3:27)

242 tests completed, 2 failed
:ratpack-manual:test FAILED
```

do these tests sometimes have spurious failures?

tednaleid commented:
ok, I now see that this is failing on master too: https://snap-ci.com/ratpack/ratpack/branch/master/logs/defaultPipeline/1451/Test?back_to=build_history

tednaleid commented:
I added a variant of `RxRatpack#fork` that takes an `Action<RegistrySpec>` to allow the injection of values into the registry of the forked execution.

I also added a version of `forkEach` that allows the same access to the registry.  I added this for parity with the `fork` code and I had also hit this as a real use-case in my production code this week.  The solution I used will be much cleaner with this option.

alkemist commented:
Solid. Great contribution @tednaleid .
